adr/ADR-43.md

```diff
@@ -0,0 +1,87 @@
+ * A stream may have a general MaxAge but some messages may have infinite retention, think a schema or type hints in a KV bucket that is forever while general keys have TTLs
+ * TODO: user use cases
+
+Related issues [#3268](https://github.com/nats-io/nats-server/issues/3268)
+
+## Per-Message TTL
+
+We will allow a message to supply a TTL as a duration in a header called `Nats-TTL` followed by the duration as 
+nanoseconds, the sever may not support nanosecond level precision and clients can choose to expose TTL as
+seconds or nanoseconds.
+
+The duration will be used by the server to calculate the deadline for removing the message based on its Stream 
+timestamp and the stated duration.
+
+The TTL may not exceed the Stream MaxAge to give administrators a hard cap for the purposes of managing resources.
+
+A TTL of zero will be ignored, any other unparsable value will result in a error reported in the Pub Ack and the message
+being discarded.
+
+Setting the header `Nats-No-Expire` to `1` will result in a message that will never be expired.
+
+## Limit Tombstones
+
+Several scenarios for server-created tombstones can be imagined, the most often requested one though is when MaxAge
+removes last value (ie. the current value) for a Key.
+
+In this case when the server removes a message and the message is the last in the subject it would place a message 
+with a TTL matching the Stream configuration value.  The following headers would be placed:
+
+```
+Nats-Applied-Limit: MaxAge
+Nats-TTL: 1000000000
+```
+
+The `Nats-Limit-Applied` field is there to support future expansion of this feature.
+
+This behaviour is off by default unless opted in on the Stream Configuration.
+
+## Publish Acknowledgements
+
+We could optionally extend the `PubAck` as follows:
+
+```golang
+type PubAck struct {
+	MsgTTL    uint64 `json:"msg_ttl,omitempty"`
+}
+```
+
+This gives clients a chance to confirm, without Stream Info or should the Stream be edited after Info, if the TTL 
+got applied.
+
+## Stream Configuration
+
+Weather or not a stream support this behavior should be a configuration opt-in. We want clients to definitely know 
+when this is supported which the opt-in approach with a boolean on the configuration would make clear.
+
+We have to assume someone will want to create a replication topology where at some point in the topology these tombstone
+type messages are retained for an audit trail. So a Stream with this feature enabled can replicate to one with it 
+disabled and all the messages that would have been TTLed will be retained.
+
+```golang
+type StreamConfig struct {
+	// AllowMsgTTL allows header initiated per-message TTLs
+	AllowMsgTTL bool          `json:"allow_msg_ttl"`
+
+	// LimitsTTL activates writing of messages when limits are applied with a specific TTL
+	LimitsTTL   time.Duration `json:"limits_ttl"`
+}
+```
```

Comment:
The above is the reason why the capability is on the server, and not a config on the stream. During the lifecycle of the client the stream may change configuration, so now the stream has a config that is different than what was read...

Reply:
I believe we'll need to support enabling and disabling this at different points in a replication chain so that some downstream source dont start expiring messages if thats not required.

I think we need to consider even removing these headers once ingested - along with a few others thats caused issues with replication - but so far there's not been much appetite for that idea and always reasons to keep them downstream.  So the most feasible is a boolean as here and this config item since its not once size fits all - though I did consider defaulting to MaxAge maybe